tests/modules/layers/test_anyprecision_optimizer.py

```diff
@@ -0,0 +1,165 @@
+            adam_opt.step()
+            anyprecision_adam.step()
+
+            # Ensure params are modified from original
+            if i == 0:
+                for p1, p2 in zip(model.parameters(), model_orig_params):
+                    assert not torch.equal(p1, p2)
+
+            for p1, p2 in zip(model.parameters(), model_clone.parameters()):
+                assert_expected(p1, p2)
+
+    def test_adam_equivalence(self, device="cpu"):
+        """
+        Tests that AnyPrecisionAdamW is equivalent to AdamW when
+        kahan summation and different dtypes for momentum, variance,
+        and compensation buffer are turned off (i.e. all float32).
+        """
+        if device == "cuda" and not torch.cuda.is_available():
+            # raise unittest.SkipTest("CUDA not available")
+            pytest.skip(reason="CUDA not available")
+
+        model = nn.Sequential(nn.Linear(5, 5), nn.Linear(5, 5), nn.Linear(5, 5))
+        if device == "cuda":
+            model.cuda()
+
+        model_clone = deepcopy(model)
+
+        self._test_adam_equivalence(model, model_clone)
+
+    def test_bfloat16_states(self, device="cpu"):
+        """verify that AnyPrecision is running using bfloat16 for states when specified (momentum, variance)"""
+        simple_model = nn.Sequential(
+            nn.Linear(5, 10, dtype=torch.bfloat16),
+            nn.GELU(),
+            nn.Linear(10, 2, dtype=torch.bfloat16),
+        )
+        simple_model.to(torch.bfloat16)
+        simple_model.to(device)
+
+        anyprecision_adam = AnyPrecisionAdamW(
+            simple_model.parameters(),
+            variance_dtype=torch.bfloat16,
+            momentum_dtype=torch.bfloat16,
+        )
+
+        for i in range(6):
+            anyprecision_adam.zero_grad()
+            inp = torch.randn(
+                5,
+                5,
+                dtype=torch.bfloat16,
+                device=next(simple_model.parameters()).device,
+            )
+            simple_model(inp).sum().backward()
+            anyprecision_adam.step()
+
+        for group in anyprecision_adam.param_groups:
+            for p in group["params"]:
+                state = anyprecision_adam.state[p]
+                assert state["exp_avg"].dtype == torch.bfloat16
+                assert state["exp_avg_sq"].dtype == torch.bfloat16
+
+    def test_kahan_summation(self, device="cpu"):
+        """verify that AnyPrecision is properly using Kahan summation when specified (momentum, variance)"""
+        simple_model = nn.Sequential(nn.Linear(5, 10), nn.GELU(), nn.Linear(10, 2))
+        simple_model.to(torch.bfloat16)
+        simple_model.to(device)
+
+        anyprecision_adam = AnyPrecisionAdamW(
+            simple_model.parameters(),
+            variance_dtype=torch.bfloat16,
+            momentum_dtype=torch.bfloat16,
+            use_kahan_summation=True,
+            compensation_buffer_dtype=torch.bfloat16,
+        )
+
+        expected_kahan_buffer = torch.tensor(
```

Comment:
how did we get this buffer?

Reply:
Yeah a short comment like "values determined by comparison to reference implementation" (assuming that's what you did) should suffice

Reply:
added comment that these are pre-computed buffers determined by comparison to reference impl.